vmdb/app/models/zone.rb

```diff
@@ -162,29 +170,33 @@ def ems_clouds
 
   def availability_zones
     MiqPreloader.preload(self.ems_clouds, :availability_zones)
-    self.ems_clouds.collect(&:availability_zones).flatten
+    ems_clouds.flat_map(&:availability_zones)
   end
 
   def vms_without_availability_zone
     MiqPreloader.preload(self, :ext_management_systems => :vms)
-    self.ext_management_systems.collect do |e|
+    ext_management_systems.flat_map do |e|
       e.kind_of?(EmsCloud) ? e.vms.select { |vm| vm.availability_zone.nil? } : []
-    end.flatten
+    end
   end
 
   def vms_and_templates
     MiqPreloader.preload(self, :ext_management_systems => :vms_and_templates)
-    self.ext_management_systems.collect(&:vms_and_templates).flatten
+    ext_management_systems.flat_map(&:vms_and_templates)
   end
 
   def vms
     MiqPreloader.preload(self, :ext_management_systems => :vms)
-    self.ext_management_systems.collect(&:vms).flatten
+    ext_management_systems.flat_map(&:vms)
+  end
+
+  def self.vms_without_a_zone
+    Vm.where(:ems_id => nil).all
```

Comment:
I responded. We cannot provide a condition for `storage_without_a_zone`, or at least I don't know how. The caller simply calls `storage_without_a_zone` without distinguishing Storage or other resources.